src/pp/if_parser.rs

```diff
@@ -0,0 +1,436 @@
+use crate::token::{Integer, PreprocessorError, Punct};
+
+use super::{Define, Location, MELexer, MacroProcessor, Step, StepExit, Token, TokenValue};
+use std::{collections::HashMap, rc::Rc, vec::IntoIter};
+
+struct IfLexer<'macros> {
+    tokens: IntoIter<Token>,
+    defines: &'macros HashMap<String, Rc<Define>>,
+}
+
+pub(super) struct IfParser<'macros> {
+    lexer: IfLexer<'macros>,
+    macro_processor: MacroProcessor,
+    location: Location,
+
+    parsing_if: bool,
+    carry: Option<Token>,
+}
+
+impl<'macros> IfParser<'macros> {
+    /// Builds a new IfParser that can be reused
+    ///
+    /// `parsing_if` indicates wether or not non defined macros should be
+    /// replaced with 0
+    pub fn new(
+        tokens: Vec<Token>,
+        defines: &'macros HashMap<String, Rc<Define>>,
+        location: Location,
+        parsing_if: bool,
+    ) -> Self {
+        IfParser {
+            lexer: IfLexer {
+                tokens: tokens.into_iter(),
+                defines,
+            },
+            macro_processor: MacroProcessor::default(),
+            location,
+
+            parsing_if,
+            carry: None,
+        }
+    }
+
+    fn raw_next(&mut self) -> Option<Token> {
+        self.carry
+            .take()
+            .or_else(|| self.macro_processor.step(&mut self.lexer).ok())
+    }
+
+    pub fn raw_peek(&mut self) -> Option<Token> {
+        self.carry.clone().or_else(|| {
+            let token = self.macro_processor.step(&mut self.lexer).ok();
+
+            self.carry = token.clone();
+
+            token
+        })
+    }
+
+    fn next(&mut self) -> Step<Option<Token>> {
+        let token = match self.raw_next() {
+            Some(t) => t,
+            None => return Ok(None),
+        };
+
+        Ok(match token.value {
+            TokenValue::Ident(ref name) if name != "defined" => {
+                match self.add_define(name, token.location)? {
+                    Some(t) => Some(t),
+                    None => self.next()?,
+                }
+            }
+            _ => Some(token),
+        })
+    }
+
+    fn peek(&mut self) -> Step<Option<Token>> {
+        let token = match self.raw_peek() {
+            Some(t) => t,
+            None => return Ok(None),
+        };
+
+        Ok(match token.value {
```

Comment:
This is very similar to `next`, maybe there is a way to put the logic for "defined" and defines only inside `next`?